theseus/theseus_layer.py

```diff
@@ -93,3 +103,88 @@ def device(self) -> torch.device:
     @property
     def dtype(self) -> torch.dtype:
         return self.objective.dtype
+
+
+def _forward(objective, optimizer, optimizer_kwargs, input_data):
+    objective.update(input_data)
+    info = optimizer.optimize(**optimizer_kwargs)
+    vars = [var.data for var in objective.optim_vars.values()]
+    return vars, info
+
+
+class TheseusLayerDLMForward(torch.autograd.Function):
+    """
+    Functionally the same as the forward method in a TheseusLayer
+    but computes the direct loss minimization in the backward pass.
+    """
+
+    @staticmethod
+    def forward(ctx, objective, optimizer, optimizer_kwargs, input_data, epsilon, *params):
+        optim_vars, info = _forward(objective, optimizer, optimizer_kwargs, input_data)
+
+        ctx.input_data = input_data.copy()
+        ctx.objective = objective
+        ctx.epsilon = epsilon
+
+        # Ideally we compute this in the backward function, but if we try to do that,
+        # it ends up in an infinite loop because it depends on the outputs of this function.
+        with torch.enable_grad():
+            grad_sol = torch.autograd.grad(objective.error_squared_norm().sum(), params, retain_graph=True)
+
+        ctx.save_for_backward(*params, *grad_sol, *optim_vars)
+        ctx.n_params = len(params)
+        return (*optim_vars, info)
+
+
+    @staticmethod
+    @once_differentiable
+    def backward(ctx, *grad_outputs):
+        saved_tensors = ctx.saved_tensors
+        params = saved_tensors[:ctx.n_params]
+        grad_sol = saved_tensors[ctx.n_params:2 * ctx.n_params]
+        optim_vars = saved_tensors[2 * ctx.n_params:]
+        grad_outputs = grad_outputs[:-1]
+
+        objective = ctx.objective
+        epsilon = ctx.epsilon
+
+        # Update the optim vars to their solutions.
+        input_data = ctx.input_data
+        values = dict(zip(objective.optim_vars.keys(), optim_vars))
+        input_data.update(values)
+
+        # Construct backward objective.
+        bwd_objective = objective.copy()
```

Comment:
Like discussed offline, we should be able to build bwd_objective once first time DLM is triggered and then have it saved for future calls.